Add async `Wifi` trait

MabezDev opened the pull request:
- Add `AsyncWifi` trait, a mirror of the `Wifi` trait
- cfg behind nightly feature

Initially, I had the idea of building these traits on top of the existing `Wifi` trait, methods like `{get|set}_configuration` could be used from the original trait, and only some methods would need to be `async`'ified. It occurred to me that not every implementation will have instant access to its configuration, for instance, for a WiFi module that communicates over SPI/SDIO/UART, getting the configuration back would indeed be blocking and thus asynchronous versions of these methods must be available.

This is just an initial idea, that for the most part works well so far. Review welcome :).

ivmarkov commented:
@MabezDev  The one problem with async-ifying the trait 1:1 as you did is that it becomes essentially a single-threaded operation, in that - unless you cancel(=`drop()`) the future - you have to wait for e.g. `set_configuration` to be finished before you can call other methods. But **I don't know** if that's a restriction in practice.

Following concrete feedback:
* Perhaps name it just `Wifi`. The name collides with the sync variant, but on the other hand, we are not using the `Async` prefix anywhere else in `embedded-svc`'s `asynch` modules
*  If you could switch to GAT + TAIT definition of the trait. We are not (yet) using the `async-fn-in-trait` nightly feature in `embedded-svc` because a) it is incomplete b) there is still a slim chance that `type-alias-impl-trait` would be stabilized soon(er) than `async-fn-in-trait`, as the former entered the FCP a month ago

MabezDev commented:
> you have to wait for e.g. set_configuration to be finished before you can call other methods

Wouldn't this be the case in the blocking trait too? `set_configuration` takes `&mut self`, so if internally it takes a while to actually set the configuration, nothing else can use it.

I've renamed the trait, I agree we should just call it `Wifi`. I've switched to GAT + TAIT in the latest commit (untested, but will test in a project tomorrow).

ivmarkov commented:
@MabezDev You need some `#[cfg(feature = "alloc")]` extra protections.

ivmarkov commented on src/wifi.rs:
```diff
@@ -412,3 +412,176 @@ where
+        where
+            Self: 'a;
+
+        type ScanFuture<'a>: Future<Output = Result<alloc::vec::Vec<AccessPointInfo>, Self::Error>>
```
`#[cfg(feature = "alloc")]`

ivmarkov commented on src/wifi.rs:
```diff
@@ -412,3 +412,176 @@ where
+        where
+            Self: 'a;
+
+        type ScanFuture<'a> = W::ScanFuture<'a>
```
Ditto.
